File: adapters/typescript/crap4ts/src/evaluate.ts
import { computeCRAP } from "./crap.js";
import { buildKey } from "./key.js";
import type { FunctionComplexity, Report } from "./types.js";

export function evaluate(
  functions: FunctionComplexity[],
  coverage: Map<string, number>,
  ceiling: number,
): Report {
  const report: Report = {
    ceiling,
    functions: [],
    summary: {
      total: functions.length,
      failing: 0,
      max_crap: 0,
    },
  };

  for (const fn of functions) {
    const key = buildKey(fn.file, fn.line, fn.func);
    const matched = coverage.has(key);
    const coverageFraction = matched ? clampCoverage(coverage.get(key) ?? 0) : 0;
    const coveragePercent = coverageFraction * 100;
    const score = computeCRAP(fn.complexity, coveragePercent);
    const pass = score <= ceiling;

    if (!pass) {
      report.summary.failing += 1;
    }
    if (score > report.summary.max_crap) {
      report.summary.max_crap = score;
    }

    report.functions.push({
      file: fn.file,
      line: fn.line,
      func: fn.func,
      complexity: fn.complexity,
      coverage: coveragePercent,
      crap: score,
      pass,
      matched,
    });
  }

  return report;
}

function clampCoverage(value: number): number {
  if (Number.isNaN(value)) {
    return 0;
  }
  if (value < 0) {
    return 0;
  }
  if (value > 1) {
    return 1;
  }
  return value;
}

File: adapters/rust/crap4rs/src/evaluate.rs
use std::collections::HashMap;

use crate::model::{build_key, FunctionComplexity, FunctionReport, Report, Summary};
use crate::score::compute_crap;

pub fn evaluate(
    functions: &[FunctionComplexity],
    coverage_percent: &HashMap<String, f64>,
    ceiling: f64,
) -> Report {
    let mut reports = Vec::with_capacity(functions.len());
    let mut failing = 0_usize;
    let mut max_crap = 0.0_f64;

    for function in functions {
        let key = build_key(&function.file, function.line, &function.func_name);
        let matched = coverage_percent.contains_key(&key);
        let coverage = coverage_percent.get(&key).copied().unwrap_or(0.0);
        let crap = compute_crap(function.complexity, coverage);
        let pass = crap <= ceiling;
        if !pass {
            failing += 1;
        }
        if crap > max_crap {
            max_crap = crap;
        }

        reports.push(FunctionReport {
            file: function.file.clone(),
            line: function.line,
            func_name: function.func_name.clone(),
            complexity: function.complexity,
            coverage,
            crap,
            pass,
            matched,
        });
    }

    Report {
        ceiling,
        summary: Summary {
            total: reports.len(),
            failing,
            max_crap,
        },
        functions: reports,
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use super::evaluate;
    use crate::model::{build_key, FunctionComplexity};

    #[test]
    fn uncovered_functions_are_scored_and_marked_unmatched() {
        let functions = vec![
            FunctionComplexity {
                file: "src/lib.rs".to_string(),
                line: 10,
                func_name: "covered".to_string(),
                complexity: 5,
            },
            FunctionComplexity {
                file: "src/lib.rs".to_string(),
                line: 30,
                func_name: "uncovered".to_string(),
                complexity: 5,
            },
        ];

        let mut coverage = HashMap::new();
        coverage.insert(build_key("src/lib.rs", 10, "covered"), 50.0);

        let report = evaluate(&functions, &coverage, 20.0);
        assert_eq!(report.summary.total, 2);

        let covered = &report.functions[0];
        assert!(covered.matched);
        assert!((covered.coverage - 50.0).abs() < 1e-9);

        let uncovered = &report.functions[1];
        assert!(!uncovered.matched);
        assert!((uncovered.coverage - 0.0).abs() < 1e-9);
        assert!((uncovered.crap - 30.0).abs() < 1e-9);
    }
}
